.pre-commit-config.yaml

```diff
@@ -0,0 +1,40 @@
+repos:
+    - repo: https://github.com/pre-commit/pre-commit-hooks
+      rev: v2.5.0
+      hooks:
+        - id: check-added-large-files
+        - id: check-merge-conflict
+        - id: check-yaml
+        - id: end-of-file-fixer
+    - repo: local
+      hooks:
+        -   id: run-file-types
+            name: Checking File Types...
+            entry: python3 tests/lint/check_file_type.py
+            language: system
+            always_run: true
+            pass_filenames: false
+        -   id: run-headers-check
+            name: Checking ASF License Headers...
+            entry: tests/lint/check_asf_header.sh --local
+            language: system
+            always_run: true
+            pass_filenames: false
+        -   id: run-black
```

Comment:
black, pylint, mypy, possibly others here that are in 'local' can be run directly from pre-commit using pre-canned configurations.  e.g.
- repo: https://github.com/psf/black
  rev: 21.5b1
  hooks:
    - id: black

Perhaps another approach here is to add those three here in precommit and replace the relevant set of tests/lint/*.sh scripts and supporting machinery and tool installs in the docker with a single tests/lint/pre-commit.sh such that pre-commit manages the versioning / installation and execution of those tools and the standard tvm CI build test environment just leverages pre-commit.   Doing this moves all the configuration of those tools to one place, it removes the need to have CI docker infrastructure manage linter versions, and means that for all the tooling fully managed by pre-commit, pre-commit can be used by devs directly on the command line or in their .git/hooks without needing to either spin up the docker environment.

Reply:
Indeed, this can work without getting involved with the docker images. 

In my case, I got the from `pip freeze` the versions that docker uses and install them locally to have the same environment. 

I agree, I am going to update the config file with pre-canned versions matching the lint environment. 

I think the goal of having such hook available is to reduce the fails in 1) Sanity checks stages and 2) the builds MacOs and Windows builds that are running the minimal unit tests.

Reply:
i see, but then we'll have two places to change when we list versions, no? also, while the repo pinning will address the version of that particular tool, i don't believe it does anything to pin the dependencies of those tools. in particular, different versions of `astroid` can produce wildly different results when combined with pylint.

Reply:
My proposal above would leave the linter tools versioned via precommit only, the CI docker environment would call pre-commit, but would not need to install or version tools directly.

Reply:
@mshawcroft after some reading, it appears that `pre-commit` will create a venv and effectively run `pip install <tool>=<version>`. i think this won't quite achieve the same thing as the docker container without specifying `additional_dependencies`--each of these named Python packages has dependencies that need to exactly match those in `ci-lint`. you could do that by cross-referencing `pip3 freeze` from ci-lint. [this blog](https://objectpartners.com/2020/06/15/consistent-python-environments-with-poetry-and-pre-commit-hooks/) discusses integrating `pre-commit` with a local venv, which goes a bit further towards this (though we are still blocked on `requirements.txt` due to CI container rebuild pipeline).

i'm okay with committing something with limitations, just noting what they are, and i would prefer we document those limitations and/or a recommended strategy to work around those somewhere. i do think that as we address the `requirements.txt` shortcoming, we can most likely resolve this issue outside docker (python version may impact this as well).

Reply:
@areusch there should be no need to provide additional constraints on their individual dependencies, if there is, that suggests an upstream bug w.r.t install_requires, are you aware of a specific issue or just concerned that there might be an issue?

I'd expect that having pre-commit specify the top level version for each lint tool is sufficient that is the normal way to use pre-commit.

Reply:
@mshawcroft so most python package dependencies are fairly loose in order to allow you to create mega-venv. imagine if a middleware required a precise version of a common library e.g. `attrs`. it would be very difficult to combine such middleware with other larger packages in the same venv while respecting the constraints of `install_requires`. therefore, many packages don't specify exact `install_requires`. you can [see](https://github.com/PyCQA/pylint/commit/fed98574f909bb34cff32b63db4e93d5271cee32#diff-fa602a8a75dc9dcc92261bac5f533c2a85e34fcceaff63b3a3a81d9acde2fc52) that when pylint updates its `astroid` dep, it merely pushes along the `>=` constraint, but now that astroid 2.6.2 is [released](https://pypi.org/project/astroid/#history), that old constraint is ambiguous.

this lax specification has made python easy to use, but as a result, API mismatches pervade the ecosystem. In general, duck typing has minimized the impact of these--up to the point that you want to reproduce exact results machine-to-machine. at this point, you need a lockfile and potentially the new pip dep resolver (note that the `install_requires` constraints were [not even enforced](https://pip.pypa.io/en/latest/user_guide/#changes-to-the-pip-dependency-resolver-in-20-3-2020) until last year). this is why lint is particularly sensitive to the exact set of packages installed in the venv, and why i don't currently recommend to use anything checked-in other than the docker container to repro ci lint results right now.

Reply:
@areusch can you share a link to a bug on the pylint project raised as a result of inadequate install_requires constraints?

Reply:
i don't have one offhand, but if you look at the default pylint bug template, they want you to run `pylint --version` and that outputs the exact versions of python, astroid, and pylint used to trigger the bug. if we are considering the case where `pylint` is the lone thing installed into a venv, the bug would have to essentially be a released regression in `astroid`. if we're considering a case where other things dependent on e.g. `astroid` were installed (especially after `pylint` was installed), then i think it'd be easier to contrive a case (i _think _ if another package locked `astroid` to an arbitrary version, pip < 20.3 could downgrade it past `pylint`'s `install_requires`).